src/lib/Libifl/pbsD_deljoblist.c

```diff
@@ -46,160 +46,471 @@
+
+	return recv_deljob(fd);
+}
+
+/**
+ * @brief	Broadcast deljob request to other servers in a multi-server set up
+ * 			This handles the cases where a job was moved to another server
  *
- * @return      int
- * @retval      DIS_SUCCESS(0)  success
- * @retval      error code      error
+ * @param[in] c - connection handle (virtual)
+ * @param[in] orig_fd - connection handle of the server which originally owned the job
+ * @param[in] function - req type
+ * @param[in] jobids - the job id list
+ * @param[in] numjids - the count of job ids
+ * @param[in] extend - extend string for req
  *
+ * @return int
+ * @retval 0 for Success
+ * @retval 1 for Error
  */
-int
-PBSD_deljoblist_put(int c, int function, char **jobids, int numofjobs, char *extend, int prot, char **msgid)
+static int
+broadcast_deljob(int c, int orig_fd, int function, char **jobids, int numids, char *extend)
 {
-	int rc;
+	int *skip_list;
+	svr_conn_t *conns = get_conn_svr_instances(c);
+	int job_found = 0;
+	int i;
+
+	if (conns == NULL) {
+		pbs_errno = PBSE_NOCONNECTS;
+		return 1;
+	}
 
-	if (prot == PROT_TCP) {
-		DIS_tcp_funcs();
-	} else {
-		if ((rc = is_compose_cmd(c, IS_CMD, msgid)) != DIS_SUCCESS)
-			return rc;
+	skip_list = calloc(get_num_servers(), sizeof(int));
+	if (skip_list == NULL) {
+		pbs_errno = PBSE_SYSTEM;
+		return 1;
 	}
 
-	if ((rc = encode_DIS_ReqHdr(c, function, pbs_current_user)) ||
-		(rc = encode_DIS_JobsList(c, jobids, numofjobs)) ||
-		(rc = encode_DIS_ReqExtend(c, extend))) {
-		if (prot == PROT_TCP) {
-			if (set_conn_errtxt(c, dis_emsg[rc]) != 0)
-				return (pbs_errno = PBSE_SYSTEM);
+	/* Broadcast it to all servers */
+	for (i = 0; i < get_num_servers(); i++) {
+		if (conns[i].sd > 0 && conns[i].sd != orig_fd) {
+			if (send_deljob(conns[i].sd, function, jobids, numids, extend))
+				skip_list[i] = 1;
+		} else
+			skip_list[i] = 1;
+	}
+	for (i = 0; i < get_num_servers(); i++) {
+		if (!skip_list[i]) {
+			struct batch_deljob_status *ret = NULL;
+
+			ret = recv_deljob(conns[i].sd);
+			if (ret == NULL || ret->code == PBSE_HISTJOBDELETED)
+				job_found = 1;
 		}
-		return (pbs_errno = PBSE_PROTOCOL);
 	}
 
-	if (prot == PROT_TPP) {
-		pbs_errno = PBSE_NONE;
-		if (dis_flush(c))
-			pbs_errno = PBSE_PROTOCOL;
-		return pbs_errno;
+	free(skip_list);
+
+	return !job_found;
+}
+
+/**
+ * @brief	Handle deleting a list of jobs in batches
+ *
+ * @param[in] c - communication handle
+ * @param[in] function - req type
+ * @param[in] jobids - the job id list
+ * @param[in] numjids - the count of job ids
+ * @param[in] extend - extend string for req
+ *
+ * @return	int
+ * @retval	0	success
+ * @retval	!0	error
+ *
+ */
+static struct batch_deljob_status *
+PBSD_deljoblist(int c, int function, char **jobids, int numjids, char *extend)
+{
+	int i;
+	svr_jobid_list_t *svr_itr = NULL;
+	struct batch_deljob_status *retlist = NULL;
+	svr_jobid_list_t *svr_jobid_list_hd = NULL;
+	int num_servers = get_num_servers();
+
+	if (num_servers > 1) {
```

Comment:
> Check the output of get_conn_svr_instances() to arrive at this decision. get_num_servers() is also not relevant for remote clusters.

Why is that? get_num_servers() should return 1 for a single server (remote or local), right? We should make these functions as generic as possible for both multi and single server setups.

Reply:
* PBSD_deljoblist() can be used to invoke with an instance fd if the client knows where the job belongs to. In that case, get_num_servers() will misguide you and say its a multi-svr cluster although you could use the single server code path.
* get_num_servers() can be made generic for both remote and local if you pass some argument to it. It should be able to distinguish what you are calling it for. For that matter get_conn_svr_instances() is already being invoked on an fd and it can return the right number of fd based on what is passed to it.
I would suggest the following to make the same decision:
```
def is_msvr(fd) {
  conns = get_conn_svr_instances(c);
  if (!conns || conns[0]->sd == fd)
          return FALSE
  return TRUE
}
```

Reply:
>  In that case, get_num_servers() will misguide you and say its a multi-svr cluster although you could use the single server code path.

Why though? get_num_servers() should return 1 in case of a single server, so a simple check for > 1 should be enough to know whether we are in single or multi-server setup, right? I think I'm missing something.

Reply:
oh wait, i think i get what you mean, even in a multi-server setup, one might call this function with a single server's fd .. let me take care of it